crates/accelerate/src/high_level_synthesis.rs

```diff
@@ -171,82 +213,767 @@ impl QubitTracker {
+
+        if synthesized_qubits.len() > input_qubits.len() {
+            let qubits_to_replace: Vec<usize> =
+                (input_qubits.len()..synthesized_qubits.len()).collect();
+            tracker.replace_state(&saved_tracker, qubits_to_replace);
         }
+
+        output_circuit_and_qubits = Some((synthesized_circuit, synthesized_qubits));
     }
 
-    /// Returns the global qubits corresponding to the given local qubit
-    fn to_global(&self, qubit: usize) -> usize {
-        self.local_to_global[qubit]
+    Ok(output_circuit_and_qubits)
+}
+
+/// Attempts to synthesize an operation using available plugins.
+///
+/// The input to this function is the operation to be synthesized and a list of global
+/// qubits over which this operation is defined.
+///
+/// The function returns either the synthesized quantum circuit and the global qubits over
+/// which it's defined, or ``None`` in the case that no synthesis methods are available or
+/// applicable (for instance, when there is an insufficient number of auxiliary qubits).
+///
+/// Internally, this function calls the Python function.
+///
+/// Currently, this function does not update the qubit tracker, which is handled upstream.
+fn synthesize_op_using_plugins(
+    py: Python,
+    data: &Bound<HighLevelSynthesisData>,
+    tracker: &mut QubitTracker,
+    input_qubits: &[usize],
+    op: &OperationRef,
+    params: &[Param],
+    label: Option<&str>,
+) -> PyResult<Option<(CircuitData, Vec<usize>)>> {
+    let mut output_circuit_and_qubits: Option<(CircuitData, Vec<usize>)> = None;
+
+    let op_py = match op {
+        OperationRef::StandardGate(standard) => {
+            standard.create_py_op(py, Some(params), label)?.into_any()
+        }
+        OperationRef::StandardInstruction(instruction) => instruction
+            .create_py_op(py, Some(params), label)?
+            .into_any(),
+        OperationRef::Gate(gate) => gate.gate.clone_ref(py),
+        OperationRef::Instruction(instruction) => instruction.instruction.clone_ref(py),
+        OperationRef::Operation(operation) => operation.operation.clone_ref(py),
+        OperationRef::Unitary(unitary) => unitary.create_py_op(py, label)?.into_any(),
+    };
+
+    let res = HLS_SYNTHESIZE_OP_USING_PLUGINS
+        .get_bound(py)
+        .call1((op_py, input_qubits, data, tracker.clone()))?
+        .extract::<Option<(QuantumCircuitData, Vec<usize>)>>()?;
+
+    if let Some((quantum_circuit_data, qubits)) = res {
+        output_circuit_and_qubits = Some((quantum_circuit_data.data, qubits));
     }
 
-    /// Returns the global qubits corresponding to the given local qubit
-    fn to_globals(&self, qubits: Vec<usize>) -> Vec<usize> {
-        qubits.iter().map(|q| self.local_to_global[*q]).collect()
+    Ok(output_circuit_and_qubits)
+}
+
+/// Synthesizes an operation.
+///
+/// This function is currently called by the default plugin for annotated operations to
+/// synthesize the base operation.
```

Comment:
It might be nice to add what the expected type of `py_op` is, on the Python side 🙂

Reply:
Done in [8a135a3](https://github.com/Qiskit/qiskit/pull/13813/commits/8a135a32a1fc7a39040922a96e357c1600d744c5)